ledger-service/http-json/src/itlib/scala/http/AbstractHttpServiceIntegrationTest.scala

```diff
@@ -774,12 +772,172 @@ abstract class QueryStoreAndAuthDependentIntegrationTest
+        VA.record(
+          Ref.Identifier assertFromString "ignored:Disclosure:CheckVisibility",
+          ShRecord(disclosed = VAx.contractIdDomain),
+        )
+
+      final case class ContractToDisclose(
+          alice: domain.Party,
+          toDiscloseCid: domain.ContractId,
+          toDisclosePayload: lav1.value.Record,
+          ctMetadata: DC.Metadata,
+      )
+
+      def contractToDisclose(
+          fixture: HttpServiceTestFixtureData,
+          junkMessage: String,
+      ): Future[ContractToDisclose] = for {
+        (alice, jwt, applicationId, _) <- fixture.getUniquePartyTokenAppIdAndAuthHeaders("Alice")
+        // we're using the ledger API for the initial create because timestamp
+        // is required in the metadata
+        toDisclosePayload = argToApi(toDiscloseVA)(ShRecord(owner = alice, junk = junkMessage))
+        createCommand = util.Commands.create(
+          refApiIdentifier(ToDisclose),
+          toDisclosePayload,
+        )
+        initialCreate = SubmitAndWaitRequest(
+          Some(
+            Commands(
+              commandId = uniqueCommandId().unwrap,
+              applicationId = applicationId.unwrap,
+              actAs = domain.Party unsubst Seq(alice),
+              commands = Seq(Command(createCommand)),
+            )
+          )
+        )
+        createResp <- fixture.client.commandServiceClient
+          .submitAndWaitForTransaction(initialCreate, Some(jwt.value))
+        (toDiscloseCid, ctMetadata) = inside(createResp.transaction) { case Some(tx) =>
+          inside(tx.events) { case Seq(lav1.event.Event(lav1.event.Event.Event.Created(ce))) =>
+            (
+              domain.ContractId(ce.contractId),
+              inside(ce.metadata map DC.Metadata.fromLedgerApi) { case Some(\/-(metadata)) =>
+                metadata
+              },
+            )
+          }
+        }
+      } yield ContractToDisclose(alice, toDiscloseCid, toDisclosePayload, ctMetadata)
+
+      "using decoded payload" in withHttpService { fixture =>
+        import fixture.encoder
+        val junkMessage = s"some test junk ${uniqueId()}"
+        for {
+          // first, set up something for alice to disclose to bob
+          ContractToDisclose(alice, toDiscloseCid, toDisclosePayload, ctMetadata) <-
+            contractToDisclose(fixture, junkMessage)
+
+          // next, onboard bob to try to interact with the disclosed contract
+          (bob, bobHeaders) <- fixture.getUniquePartyAndAuthHeaders("Bob")
+          viewportCid <- postCreateCommand(
+            domain
+              .CreateCommand(
+                TpId.Disclosure.Viewport,
+                argToApi(viewportVA)(ShRecord(owner = bob)),
+                None,
+              ),
+            fixture,
+            bobHeaders,
+          ) map resultContractId
+
+          // exercise CheckVisibility with different disclosure options
+          checkVisibility = { disclosure: Option[DC.Arguments[lav1.value.Value]] =>
```

Comment:
It might be good to have an example where there are multiple on disclosed contracts.

Reply:
I've added this to #14260.